pkg/terraform/workspace_test.go

```diff
@@ -0,0 +1,454 @@
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			r, err := tc.w.Apply(context.TODO())
+			if diff := cmp.Diff(tc.want.err, err, test.EquateErrors()); diff != "" {
+				t.Errorf("\n%s\nApply(...): -want error, +got error:\n%s", name, diff)
+			}
+			if diff := cmp.Diff(tc.want.r, r, test.EquateErrors()); diff != "" {
+				t.Errorf("\n%s\nApply(...): -want error, +got error:\n%s", name, diff)
+			}
+		})
+	}
+}
+
+func TestWorkspaceDestroy(t *testing.T) {
+	type args struct {
+		w *Workspace
+	}
+	type want struct {
+		err error
+	}
+
+	cases := map[string]struct {
+		args
+		want
+	}{
+		"Running": {
+			args: args{
+				w: NewWorkspace(directory, WithLastOperation(&Operation{Type: testType, startTime: timeNow(), endTime: nil})),
+			},
+			want: want{
+				err: errors.Errorf("%s operation that started at %s is still running", testType, timeNow().String()),
+			},
+		},
+		"Success": {
+			args: args{
+				w: NewWorkspace(
+					directory, WithExec(&testingexec.FakeExec{DisableScripts: true}), WithLogger(newLogger())),
+			},
+			want: want{},
+		},
+		"Failure": {
+			args: args{
+				w: newFailureWorkspace(),
+			},
+			want: want{
+				err: tferrors.NewDestroyFailed([]byte(errBoom.Error())),
+			},
+		},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			err := tc.w.Destroy(context.TODO())
+			if diff := cmp.Diff(tc.want.err, err, test.EquateErrors()); diff != "" {
+				t.Errorf("\n%s\nDestroy(...): -want error, +got error:\n%s", name, diff)
+			}
+		})
+	}
+}
+
+func TestWorkspaceRefresh(t *testing.T) {
+	type args struct {
+		w *Workspace
+	}
+	type want struct {
+		r   RefreshResult
+		err error
+	}
+
+	cases := map[string]struct {
+		args
+		want
+	}{
+		"Running": {
+			args: args{
+				w: NewWorkspace(directory, WithLastOperation(&Operation{Type: applyType, startTime: timeNow(), endTime: nil})),
+			},
+			want: want{
+				r: RefreshResult{
+					IsApplying: true,
+				},
+			},
+		},
+		"Success": {
+			args: args{
+				w: NewWorkspace(
+					directory, WithExec(&testingexec.FakeExec{DisableScripts: true}), WithLogger(newLogger())),
+			},
+			want: want{
+				r: RefreshResult{
+					State: state,
+				},
+			},
+		},
+		"Failure": {
+			args: args{
+				w: newFailureWorkspace(),
+			},
+			want: want{
+				err: tferrors.NewRefreshFailed([]byte(errBoom.Error())),
+			},
+		},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			r, err := tc.w.Refresh(context.TODO())
+			if diff := cmp.Diff(tc.want.err, err, test.EquateErrors()); diff != "" {
+				t.Errorf("\n%s\nRefresh(...): -want error, +got error:\n%s", name, diff)
+			}
+			if diff := cmp.Diff(tc.want.r, r, test.EquateErrors()); diff != "" {
+				t.Errorf("\n%s\nRefresh(...): -want error, +got error:\n%s", name, diff)
+			}
+		})
+	}
+}
+
+func TestWorkspacePlan(t *testing.T) {
+	type args struct {
+		w *Workspace
+	}
+	type want struct {
+		r   PlanResult
+		err error
+	}
+
+	cases := map[string]struct {
```

Comment:
A failure case for `Plan` would be useful. I'd suggest running this file's tests with `Run with Coverage` feature of Goland or IDE you have so that it highlights which lines aren't covered.

Reply:
Added. Thanks for your suggestion!